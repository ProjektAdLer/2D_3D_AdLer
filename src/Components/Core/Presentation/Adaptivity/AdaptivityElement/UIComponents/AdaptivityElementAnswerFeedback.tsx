import { useEffect, useRef } from "react";
import { useTranslation } from "react-i18next";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import {
  AdaptivityQuestion,
  AdaptivityTask,
} from "../AdaptivityElementViewModel";

export default function AdaptivityElementAnswerFeedback({
  isCorrect,
  currentQuestion,
  currentTask,
  setHeaderText,
  closeFeedback,
}: {
  isCorrect: boolean;
  currentQuestion: AdaptivityQuestion;
  currentTask: AdaptivityTask;
  setHeaderText: (text: string) => void;
  closeFeedback: () => void;
}) {
  const { t: translate } = useTranslation("learningElement");
  useEffect(() => {
    setHeaderText(
      isCorrect ? translate("rightAnswer") : translate("wrongAnswer")
    );
  }, [isCorrect, setHeaderText, translate]);
  const difficultyHint = useRef(null as string | null);
  useEffect(() => {
    currentTask.questions.some((question) => {
      return (
        question.difficulty.valueOf() > currentQuestion.difficulty.valueOf()
      );
    }) &&
      isCorrect &&
      (difficultyHint.current = translate(
        "questionSolvedWithHigherDifficultyQuestionAvailable"
      ).toString());

    currentTask.questions.some((question) => {
      return (
        question.difficulty.valueOf() < currentQuestion.difficulty.valueOf()
      );
    }) &&
      !isCorrect &&
      (difficultyHint.current = translate(
        "questionSolvedWithLowerDifficultyQuestionAvailable"
      ));
  }, [currentQuestion, currentTask, isCorrect, translate]);
  const solvedHint = useRef(null as string | null);
  useEffect(() => {
    if (
      // hint is given if the solved question is required or if a question of lower difficulty is required
      (currentTask.questions.some((question) => {
        return (
          question.isRequired &&
          question.difficulty.valueOf() < currentQuestion.difficulty.valueOf()
        );
      }) ||
        currentQuestion.isRequired) &&
      isCorrect
    ) {
      solvedHint.current = translate("requiredQuestionSolvedCorrectly");
    }
  }, [
    currentQuestion.difficulty,
    currentQuestion.isRequired,
    currentTask,
    isCorrect,
    translate,
  ]);

  return (
    <div className="flex flex-col gap-4 pl-4 pr-2 my-4 h-fit">
      <div className="flex flex-col items-start justify-start bg-buttonbgblue rounded-xl">
        {solvedHint.current && <div className="p-2">{solvedHint.current}</div>}
        {difficultyHint.current && (
          <div className="p-2">{difficultyHint.current}</div>
        )}
      </div>
      <div className="flex justify-end w-full">
        <StyledButton shape="freefloatcenter" onClick={closeFeedback}>
          <p className="text-sm">{translate("nextButton")}</p>
        </StyledButton>
      </div>
    </div>
  );
}
