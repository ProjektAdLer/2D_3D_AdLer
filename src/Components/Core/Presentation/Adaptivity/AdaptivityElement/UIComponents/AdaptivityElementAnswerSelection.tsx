import { useCallback, useEffect, useState } from "react";
import { AdaptivityQuestion } from "../AdaptivityElementViewModel";
import StyledButton, {
  StyledButtonColor,
} from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import { useTranslation } from "react-i18next";

function getIsAnyAnswerSelected(question: AdaptivityQuestion): boolean {
  return question.questionAnswers.some((a) => a.isSelected === true);
}

export default function AdaptivityElementAnswerSelection({
  question,
  submitSelection,
  closeSelection,
  setHeaderText,
}: {
  question: AdaptivityQuestion;
  submitSelection: () => void;
  closeSelection: () => void;
  setHeaderText: (headerText: string) => void;
}) {
  const [answerColors, setAnswerColors] = useState<StyledButtonColor[]>([]);
  const [isAnyAnswerSelected, setIsAnyAnswerSelected] = useState<boolean>(
    getIsAnyAnswerSelected(question)
  );

  const { t: translate } = useTranslation("learningElement");
  const [text] = useState<string>(
    (question.isMultipleChoice
      ? translate("multipleChoiceSelected")
      : translate("singleChoiceSelected")) + question.questionText
  );

  useEffect(() => {
    setHeaderText(text);
    setAnswerColors(question.questionAnswers.map(() => "default"));
  }, [setHeaderText, question.questionAnswers, text]);

  const onAnswerClicked = useCallback(
    (index: number) => {
      const newColors = question.questionAnswers.map((answer, i) => {
        if (index === i) {
          // switch selection status of clicked answer
          answer.isSelected = !answer.isSelected;
        } else if (!question.isMultipleChoice) {
          // deselect all other answers if question is single choice
          answer.isSelected = false;
        }

        return answer.isSelected ? "highlight" : "default";
      });

      setIsAnyAnswerSelected(getIsAnyAnswerSelected(question));

      setAnswerColors(newColors);
    },
    [question]
  );

  return (
    <div className="flex flex-col w-full p-2 m-auto lg:justify-between lg:flex-row">
      <div className="flex flex-wrap justify-start gap-4">
        {question.questionAnswers.map((answer, index) => (
          <StyledButton
            key={answer.answerIndex}
            shape="freefloatcenter"
            onClick={() => {
              onAnswerClicked(index);
            }}
            color={answerColors[index]}
          >
            <p className="text-sm">{answer.answerText}</p>
          </StyledButton>
        ))}
      </div>

      <div className="flex items-end justify-end gap-2 pt-8 font-bold lg:w-1/6">
        <StyledButton
          className="box-border "
          shape="freefloatcenter"
          onClick={() => {
            submitSelection();
          }}
          disabled={!isAnyAnswerSelected}
        >
          <p className="text-sm">
            {question.isMultipleChoice
              ? translate("submitAnswers")
              : translate("submitAnswer")}
          </p>
        </StyledButton>
      </div>
    </div>
  );
}
