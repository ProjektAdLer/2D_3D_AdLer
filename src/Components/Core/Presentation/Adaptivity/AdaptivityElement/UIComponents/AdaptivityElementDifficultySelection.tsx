import { useEffect, useState } from "react";
import { AdaptivityTask } from "../AdaptivityElementViewModel";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import { AdaptivityElementQuestionDifficultyTypes } from "src/Components/Core/Domain/Types/Adaptivity/AdaptivityElementQuestionDifficultyTypes";

export default function AdaptivityElementDifficultySelection({
  selectedTask,
  setHeaderText,
  onSelectDifficulty,
}: {
  selectedTask: AdaptivityTask;
  setHeaderText: (headerText: string) => void;
  onSelectDifficulty: (taskID: number) => void;
}) {
  const [difficultyButtons, setDifficultyButtons] = useState<JSX.Element[]>([]);

  useEffect(() => {
    setHeaderText(
      `Du hast Aufgabengebiet ${selectedTask.taskTitle} ausgewählt. Wähle jetzt eine Schwierigkeit aus.`
    );
  }, [setHeaderText]);

  useEffect(() => {
    setDifficultyButtons(
      selectedTask.questions.map((question) => {
        return (
          <div key={question.questionID} className="flex flex-col w-[33%] ">
            <StyledButton
              shape="freefloatcenter"
              className="w-full max-w-2xl"
              onClick={() => onSelectDifficulty(question.questionID)}
            >
              {question.difficulty ===
                AdaptivityElementQuestionDifficultyTypes.easy && "Leicht"}
              {question.difficulty ===
                AdaptivityElementQuestionDifficultyTypes.medium && "Medium"}
              {question.difficulty ===
                AdaptivityElementQuestionDifficultyTypes.hard && "Schwer"}
            </StyledButton>
            {/* TODO: Add Hinweis Schranke ja nein */}
            <StyledButton shape="freefloatcenter" className="w-full max-w-2xl">
              Hinweis
            </StyledButton>
          </div>
        );
      })
    );
  }, [selectedTask]);

  return (
    <div className="flex flex-row w-[100%] gap-2">{difficultyButtons}</div>
  );
}
