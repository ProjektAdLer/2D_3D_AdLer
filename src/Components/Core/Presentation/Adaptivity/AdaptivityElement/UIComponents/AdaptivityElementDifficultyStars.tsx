import tailwindMerge from "../../../Utils/TailwindMerge";

import requiredSolvedIcon from "../../../../../../Assets/icons/40-difficulties-adaptivity/diffculties-adaptivity-required-solved-icon.svg";
import requiredUnsolvedIcon from "../../../../../../Assets/icons/40-difficulties-adaptivity/diffculties-adaptivity-required-unsolved-icon.svg";
import notRequiredSolvedIcon from "../../../../../../Assets/icons/40-difficulties-adaptivity/diffculties-adaptivity-solved-icon.svg";
import notRequiredUnsolvedIcon from "../../../../../../Assets/icons/40-difficulties-adaptivity/diffculties-adaptivity-unsolved-icon.svg";
import placeholderIcon from "../../../../../../Assets/icons/40-difficulties-adaptivity/diffculties-adaptivity-placeholder.svg";

export enum AdaptivityElementDifficultyStarState {
  Empty,
  RequiredSolved,
  RequiredTried,
  RequiredUnsolved,
  NotRequiredSolved,
  NotRequiredTried,
  NotRequiredUnsolved,
}

enum AdaptivityElementQuestionDifficulty {
  Easy,
  Medium,
  Hard,
}

// TODO: add more star icons here
function getIcon(
  difficulty: AdaptivityElementQuestionDifficulty,
  status: AdaptivityElementDifficultyStarState
): string | undefined {
  switch (difficulty) {
    case AdaptivityElementQuestionDifficulty.Easy:
      switch (status) {
        case AdaptivityElementDifficultyStarState.RequiredSolved:
          return requiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredUnsolved:
          return requiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredTried:
          return "";
        case AdaptivityElementDifficultyStarState.NotRequiredSolved:
          return notRequiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredUnsolved:
          return notRequiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredTried:
          return "";
        default:
          return undefined;
      }
    case AdaptivityElementQuestionDifficulty.Medium:
      switch (status) {
        case AdaptivityElementDifficultyStarState.RequiredSolved:
          return requiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredUnsolved:
          return requiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredTried:
          return "";
        case AdaptivityElementDifficultyStarState.NotRequiredSolved:
          return notRequiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredUnsolved:
          return notRequiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredTried:
          return "";
        default:
          return undefined;
      }
    case AdaptivityElementQuestionDifficulty.Hard:
      switch (status) {
        case AdaptivityElementDifficultyStarState.RequiredSolved:
          return requiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredUnsolved:
          return requiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.RequiredTried:
          return "";
        case AdaptivityElementDifficultyStarState.NotRequiredSolved:
          return notRequiredSolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredUnsolved:
          return notRequiredUnsolvedIcon;
        case AdaptivityElementDifficultyStarState.NotRequiredTried:
          return "";
        default:
          return undefined;
      }
  }
}

function EmptyStarSlot({ className }: { className: string }): JSX.Element {
  return (
    /*<div
      data-testid="empty-star-slot"
      className={tailwindMerge(`${className}`)}
    />*/
    <img
      className={tailwindMerge(`${className}`) + "opacity-60"}
      alt="Platzhalter-Icon"
      src={placeholderIcon}
    />
  );
}

export default function AdaptivityElementDifficultyStars({
  easyState = AdaptivityElementDifficultyStarState.Empty,
  mediumState = AdaptivityElementDifficultyStarState.Empty,
  hardState = AdaptivityElementDifficultyStarState.Empty,
  starClassName,
  withEmptySlots = true,
}: {
  easyState?: AdaptivityElementDifficultyStarState;
  mediumState?: AdaptivityElementDifficultyStarState;
  hardState?: AdaptivityElementDifficultyStarState;
  starClassName: string;
  withEmptySlots?: boolean;
}) {
  return (
    <div className="flex items-center h-full">
      {easyState !== AdaptivityElementDifficultyStarState.Empty ? (
        <img
          className={tailwindMerge(`${starClassName}`)}
          alt=""
          src={getIcon(AdaptivityElementQuestionDifficulty.Easy, easyState)}
        />
      ) : (
        withEmptySlots && <EmptyStarSlot className={starClassName} />
      )}
      {mediumState !== AdaptivityElementDifficultyStarState.Empty ? (
        <img
          className={tailwindMerge(`${starClassName}`)}
          alt=""
          src={getIcon(AdaptivityElementQuestionDifficulty.Medium, mediumState)}
        />
      ) : (
        withEmptySlots && <EmptyStarSlot className={starClassName} />
      )}
      {hardState !== AdaptivityElementDifficultyStarState.Empty ? (
        <img
          className={tailwindMerge(`${starClassName}`)}
          alt=""
          src={getIcon(AdaptivityElementQuestionDifficulty.Hard, hardState)}
        />
      ) : (
        withEmptySlots && <EmptyStarSlot className={starClassName} />
      )}
    </div>
  );
}
