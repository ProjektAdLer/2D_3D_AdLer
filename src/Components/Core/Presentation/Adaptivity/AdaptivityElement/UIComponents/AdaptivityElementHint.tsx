import { AdaptivityElementActionTypes } from "src/Components/Core/Domain/Types/Adaptivity/AdaptivityElementActionTypes";
import { AdaptivityHint } from "../AdaptivityElementViewModel";
import { useEffect } from "react";
import { useTranslation } from "react-i18next";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";

export default function AdaptivityElementHint({
  hint,
  setHeaderText,
  closeHint,
}: {
  hint: AdaptivityHint;
  setHeaderText: (headerText: string) => void;
  closeHint: () => void;
}) {
  const { t: translate } = useTranslation("learningElement");

  useEffect(() => {
    setHeaderText(translate("hintTitle"));
  }, [setHeaderText, translate]);

  return (
    <>
      {hint.hintAction.hintActionType ===
        AdaptivityElementActionTypes.CommentAction &&
        hint.hintAction.textData !== undefined && (
          <div className="flex items-center justify-center p-2 pl-4 my-4 bg-buttonbgblue rounded-xl">
            {hint.hintAction.textData}
          </div>
        )}

      {hint.hintAction.hintActionType ===
        AdaptivityElementActionTypes.ContentAction && null}

      {hint.hintAction.hintActionType ===
        AdaptivityElementActionTypes.ReferenceAction &&
        hint.hintAction.textData !== undefined && (
          <div>{hint.hintAction.textData}</div>
        )}
      <div className="flex justify-end w-full">
        <StyledButton shape="freefloatcenter" onClick={closeHint}>
          <p className="text-sm">{translate("nextButton")}</p>
        </StyledButton>
      </div>
    </>
  );
}
