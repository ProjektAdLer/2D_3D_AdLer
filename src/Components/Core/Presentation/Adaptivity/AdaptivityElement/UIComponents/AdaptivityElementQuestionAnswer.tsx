import { useCallback, useEffect, useState } from "react";
import {
  AdaptivityAnswer,
  AdaptivityQuestion,
} from "../AdaptivityElementViewModel";
import StyledButton, {
  StyledButtonColor,
} from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";

export default function AdaptivityElementQuestionAnswer({
  question,
  submitSelection,
  setHeaderText,
}: {
  question: AdaptivityQuestion;
  submitSelection: () => void;
  setHeaderText: (headerText: string) => void;
}) {
  const [answerColors, setAnswerColors] = useState<StyledButtonColor[]>([]);

  useEffect(() => {
    setHeaderText(question.questionText);
    setAnswerColors(question.questionAnswers.map(() => "default"));
  }, [setHeaderText, question]);

  const onAnswerClicked = useCallback(
    (index: number) => {
      const newColors = question.questionAnswers.map((answer, i) => {
        if (index === i) {
          // switch selection status of clicked answer
          answer.isSelected = !answer.isSelected;
        } else {
          // deselect all other answers if question is single choice
          if (!question.isMultipleChoice) answer.isSelected = false;
        }

        return answer.isSelected ? "highlight" : "default";
      });

      setAnswerColors(newColors);
    },
    [answerColors]
  );

  return (
    <div className="flex flex-col w-full m-auto p-2">
      <div className="flex flex-col lg:grid lg:grid-cols-2 flex-wrap justify-start gap-4 ">
        {question.questionAnswers.map((answer, index) => (
          <StyledButton
            key={answer.answerIndex}
            shape="freefloatcenter"
            onClick={() => {
              onAnswerClicked(index);
            }}
            color={answerColors[index]}
          >
            {answer.answerText}
          </StyledButton>
        ))}
      </div>

      <div className="flex items-end justify-end w-full pt-8 font-bold">
        <StyledButton
          className="box-border"
          shape="freefloatcenter"
          onClick={submitSelection}
        >
          {"Antworten abgeben"}
        </StyledButton>
      </div>
    </div>
  );
}
