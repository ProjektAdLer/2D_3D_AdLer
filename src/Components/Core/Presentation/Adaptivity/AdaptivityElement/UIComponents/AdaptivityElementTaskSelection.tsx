import { useEffect, useState } from "react";
import {
  AdaptivityQuestion,
  AdaptivityTask,
} from "../AdaptivityElementViewModel";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import AdaptivityElementDifficultyStars, {
  AdaptivityElementDifficultyStarState,
} from "./AdaptivityElementDifficultyStars";
import { AdaptivityElementQuestionDifficultyTypes } from "src/Components/Core/Domain/Types/Adaptivity/AdaptivityElementQuestionDifficultyTypes";
import { useTranslation } from "react-i18next";

import requiredTaskIcon from "../../../../../../Assets/icons/required.svg";
import solvedTaskIcon from "../../../../../../Assets/icons/check-solution.svg";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";
import Observable from "src/Lib/Observable";

export function getAdaptivityQuestionStarState(
  question: AdaptivityQuestion | undefined,
): AdaptivityElementDifficultyStarState {
  if (question === undefined) return AdaptivityElementDifficultyStarState.Empty;
  else if (question.isRequired) {
    if (question.isCompleted === true)
      return AdaptivityElementDifficultyStarState.RequiredSolved;
    else if (question.isCompleted === false)
      return AdaptivityElementDifficultyStarState.RequiredTried;
    else return AdaptivityElementDifficultyStarState.RequiredUnsolved;
  } else {
    if (question.isCompleted === true)
      return AdaptivityElementDifficultyStarState.NotRequiredSolved;
    else if (question.isCompleted === false)
      return AdaptivityElementDifficultyStarState.NotRequiredTried;
    else return AdaptivityElementDifficultyStarState.NotRequiredUnsolved;
  }
}

export default function AdaptivityElementTaskSelection({
  tasks,
  reset,
  setHeaderText,
  onSelectTask,
}: Readonly<{
  tasks: AdaptivityTask[];
  reset: Observable<boolean>;
  setHeaderText: (headerText: string) => void;
  onSelectTask: (selectedTask: AdaptivityTask) => void;
}>) {
  const [taskButtons, setTaskButtons] = useState<JSX.Element[]>([]);
  const [resetting] = useObservable<boolean>(reset);

  const { t: translate } = useTranslation("learningElement");
  useEffect(() => {
    setHeaderText(translate("adaptivityIntro"));
  }, [setHeaderText, translate]);

  const taskToolTip = translate("taskToolTip").toString();
  const taskRequiredToolTip = translate("legendTaskRequired").toString();
  const taskSolvedToolTip = translate("legendTaskRequiredSolved").toString();

  useEffect(() => {
    setTaskButtons(
      tasks.map((task, i) => {
        const easyStatus = getAdaptivityQuestionStarState(
          task.questions.find(
            (question) =>
              question.difficulty ===
              AdaptivityElementQuestionDifficultyTypes.easy,
          ),
        );
        const mediumStatus = getAdaptivityQuestionStarState(
          task.questions.find(
            (question) =>
              question.difficulty ===
              AdaptivityElementQuestionDifficultyTypes.medium,
          ),
        );
        const hardStatus = getAdaptivityQuestionStarState(
          task.questions.find(
            (question) =>
              question.difficulty ===
              AdaptivityElementQuestionDifficultyTypes.hard,
          ),
        );

        return (
          <div
            key={i.toString() + "_" + task.taskID}
            className="flex flex-col text-sm"
          >
            <StyledButton
              shape="freeFloatCenter"
              className="w-full max-w-2xl"
              onClick={() => onSelectTask(task)}
              title={taskToolTip}
              data-testid={`adaptivity-task-${task.taskID}`}
            >
              <div className="align-center grid h-full w-full grid-cols-7 items-center justify-between">
                <AdaptivityElementDifficultyStars
                  easyState={easyStatus}
                  mediumState={mediumStatus}
                  hardState={hardStatus}
                  starClassName="w-4 h-4 sm:w-6 sm:h-6 "
                />

                <p className="col-span-5 pl-4 text-start text-sm">
                  {task.taskTitle}
                </p>

                <div className="flex justify-end">
                  {task.isRequired && !task.isCompleted && (
                    <img
                      alt=""
                      className={"h-6 pl-4 lg:h-8 xl:pl-8"}
                      src={requiredTaskIcon}
                      title={taskRequiredToolTip}
                      data-testid={`adaptivity-star-required-notcompleted-${task.taskID}`}
                    />
                  )}
                  {task.isRequired && task.isCompleted && (
                    <img
                      alt=""
                      className={"h-6 pl-4 lg:h-8 xl:pl-8"}
                      src={solvedTaskIcon}
                      title={taskSolvedToolTip}
                      data-testid={`adaptivity-star-required-completed-${task.taskID}`}
                    />
                  )}
                  {!task.isRequired && <div className="w-16"></div>}
                </div>
              </div>
            </StyledButton>
          </div>
        );
      }),
    );
  }, [
    tasks,
    onSelectTask,
    resetting,
    taskSolvedToolTip,
    taskRequiredToolTip,
    taskToolTip,
  ]);

  return (
    <div className="grid justify-center gap-4 gap-x-8 lg:grid-cols-2">
      {taskButtons}
    </div>
  );
}
