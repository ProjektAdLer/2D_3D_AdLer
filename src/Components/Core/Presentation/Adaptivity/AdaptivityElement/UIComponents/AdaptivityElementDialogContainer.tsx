import { useState } from "react";
import useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";
import AdaptivityElementViewModel from "../AdaptivityElementViewModel";
import IAdaptivityElementController from "../IAdaptivityElementController";
import BUILDER_TYPES from "~DependencyInjection/Builders/BUILDER_TYPES";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";
import StyledContainer from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledContainer";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import tailwindMerge from "../../../Utils/TailwindMerge";
import { AdLerUIComponent } from "../../../../Types/ReactTypes";

import quizBackgroundVRGuy from "../../../../../../Assets/misc/quizBackgrounds/vr-guy-quiz-background.png";
import AdaptivityElementTaskSelection from "./AdaptivityElementTaskSelection";

export default function AdaptivityElementDialogContainer({
  className,
}: AdLerUIComponent) {
  const [viewmodel, controller] = useBuilder<
    AdaptivityElementViewModel,
    IAdaptivityElementController
  >(BUILDER_TYPES.IAdaptivityElementBuilder);

  const [isOpen] = useObservable<boolean>(viewmodel?.isOpen);
  const [contentData] = useObservable(viewmodel?.contentData);

  // TODO: remove DEBUG data
  const [footer] = useState("Adaptivit√§tselement");

  if (!viewmodel || !controller) return null;
  if (!isOpen || !contentData) return null;

  return (
    <>
      <img
        className="absolute top-0 right-0 z-10 max-h-[93vh] w-fit max-w-[90vw] lg:max-w-[99vw]"
        alt="LearningImage!"
        src={quizBackgroundVRGuy}
      ></img>
      <StyledContainer className={tailwindMerge(className, "")}>
        <div className="fixed top-0 bottom-0 left-0 right-0 flex flex-col items-center justify-center h-full">
          <div className="p-2 rounded-lg bg-gradient-to-br from-adlerbggradientfrom to-adlerbggradientto max-w-[95%] max-h-[95%] overflow-auto">
            <div className="flex items-center justify-center w-full gap-2 p-1 pb-3 text-xl font-bold text-adlerdarkblue lg:roboto-black lg:text-2xl h-fit">
              <div className="w-full">{contentData.elementName}</div>

              <StyledButton
                onClick={controller.closeModal}
                className="w-8 h-8 p-1 text-xs roboto-black xl:w-10 xl:h-10 lg:w-10 lg:h-10 md:w-10 md:h-10 sm:w-10 sm:h-10"
              >
                X
              </StyledButton>
            </div>

            <div className="px-1 rounded-lg font-regular h-fit">
              <AdaptivityElementTaskSelection
                tasks={contentData.tasks}
                onSelectTask={controller.selectTask}
              />
            </div>

            {
              <div className="modal-footer">
                <p>{footer}</p>
              </div>
            }
          </div>
        </div>
      </StyledContainer>
    </>
  );
}
