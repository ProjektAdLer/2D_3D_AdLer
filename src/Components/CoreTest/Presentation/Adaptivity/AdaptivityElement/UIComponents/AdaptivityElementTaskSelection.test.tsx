import { render } from "@testing-library/react";
import AdaptivityElementTaskSelection, {
  getAdaptivityQuestionStarState,
} from "../../../../../Core/Presentation/Adaptivity/AdaptivityElement/UIComponents/AdaptivityElementTaskSelection";
import React from "react";
import { AdaptivityElementDifficultyStarState } from "../../../../../Core/Presentation/Adaptivity/AdaptivityElement/UIComponents/AdaptivityElementDifficultyStars";

describe("AdaptivityElementTaskSelection", () => {
  test("should render", () => {
    const { container } = render(
      <AdaptivityElementTaskSelection
        tasks={[
          {
            taskID: 0,
            taskTitle: "testTaskTitle",
            isCompleted: false,
            isRequired: false,
            requiredDifficulty: 0,
            questions: [
              {
                questionID: 0,
                questionText: "testQuestionText",
                isMultipleChoice: false,
                difficulty: 0,
                isCompleted: false,
                isRequired: false,
                hints: [],
                questionAnswers: [
                  {
                    answerIndex: 0,
                    answerText: "testAnswerText",
                    isSelected: false,
                  },
                ],
              },
            ],
          },
        ]}
        setHeaderText={() => {}}
        onSelectTask={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("task button calls onSelectTask callback when clicked", () => {
    const onSelectTaskMock = jest.fn();
    const { getByText } = render(
      <AdaptivityElementTaskSelection
        tasks={[
          {
            taskID: 42,
            taskTitle: "testTaskTitle",
            isCompleted: false,
            isRequired: false,
            requiredDifficulty: 0,
            questions: [
              {
                questionID: 0,
                questionText: "testQuestionText",
                isMultipleChoice: false,
                difficulty: 0,
                isCompleted: false,
                isRequired: false,
                hints: [],
                questionAnswers: [
                  {
                    answerIndex: 0,
                    answerText: "testAnswerText",
                    isSelected: false,
                  },
                ],
              },
            ],
          },
        ]}
        setHeaderText={() => {}}
        onSelectTask={onSelectTaskMock}
      />
    );

    getByText("testTaskTitle").click();

    expect(onSelectTaskMock).toHaveBeenCalledTimes(1);
    expect(onSelectTaskMock).toHaveBeenCalledWith(42);
  });

  test("getAdaptivityQuestionStarState returns state Empty for undefined question", () => {
    expect(getAdaptivityQuestionStarState(undefined)).toBe(
      AdaptivityElementDifficultyStarState.Empty
    );
  });

  test.each([
    {
      isRequired: true,
      isCompleted: true,
      expected: AdaptivityElementDifficultyStarState.RequiredSolved,
    },
    {
      isRequired: true,
      isCompleted: false,
      expected: AdaptivityElementDifficultyStarState.RequiredTried,
    },
    {
      isRequired: true,
      isCompleted: null,
      expected: AdaptivityElementDifficultyStarState.RequiredUnsolved,
    },
    {
      isRequired: false,
      isCompleted: true,
      expected: AdaptivityElementDifficultyStarState.NotRequiredSolved,
    },
    {
      isRequired: false,
      isCompleted: false,
      expected: AdaptivityElementDifficultyStarState.NotRequiredTried,
    },
    {
      isRequired: false,
      isCompleted: null,
      expected: AdaptivityElementDifficultyStarState.NotRequiredUnsolved,
    },
  ])(
    "getAdaptivityQuestionStarState returns state $expected for question with isRequired=$isRequired and isCompleted=$isCompleted",
    ({ isRequired, isCompleted, expected }) => {
      const result = getAdaptivityQuestionStarState({
        questionID: 0,
        questionText: "testQuestionText",
        isMultipleChoice: false,
        difficulty: 0,
        isCompleted,
        isRequired,
        hints: [],
        questionAnswers: [
          {
            answerIndex: 0,
            answerText: "testAnswerText",
            isSelected: false,
          },
        ],
      });

      expect(result).toBe(expected);
    }
  );
});
