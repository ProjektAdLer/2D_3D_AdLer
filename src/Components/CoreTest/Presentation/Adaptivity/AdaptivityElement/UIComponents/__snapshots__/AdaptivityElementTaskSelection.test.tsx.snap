// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdaptivityElementTaskSelection should render 1`] = `
<div>
  <div
    class="grid justify-center gap-4 lg:grid-cols-2 gap-x-8"
  >
    <div
      class="flex flex-col text-sm"
    >
      <div
        class=""
      >
        <button
          class="w-full max-w-2xl px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer landscape:hover:bg-adleryellow lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent active:bg-adleryellow text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"
        >
          <div
            class="grid items-center justify-between w-full h-full grid-cols-7 align-center"
          >
            <div
              class="flex items-center h-full"
            >
              <img
                alt=""
                class="w-4 h-4 sm:w-6 sm:h-6"
                src="diffculties-adaptivity-easy-tried.svg"
              />
              <img
                alt="Platzhalter-Icon"
                class="w-4 h-4 sm:w-6 sm:h-6 opacity-60"
                src="diffculties-adaptivity-placeholder.svg"
              />
              <img
                alt="Platzhalter-Icon"
                class="w-4 h-4 sm:w-6 sm:h-6 opacity-60"
                src="diffculties-adaptivity-placeholder.svg"
              />
            </div>
            <p
              class="col-span-5 pl-4 text-sm text-start"
            >
              testTaskTitle
            </p>
            <div
              class="flex justify-end"
            >
              <div
                class="w-16"
              />
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
